akka-cluster-sharding/src/test/scala/akka/cluster/sharding/SupervisionSpec.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding
+
+import akka.actor.{ Actor, ActorLogging, ActorRef, PoisonPill, Props }
+import akka.cluster.Cluster
+import akka.cluster.sharding.ShardRegion.Passivate
+import akka.pattern.{ Backoff, BackoffSupervisor }
+import akka.testkit.{ AkkaSpec, ImplicitSender }
+import com.typesafe.config.ConfigFactory
+
+import scala.concurrent.duration._
+
+object SupervisionSpec {
+  val config =
+    ConfigFactory.parseString(
+      """
+    akka.actor.provider = "cluster"
+    akka.loglevel = DEBUG
+    """)
+
+  case class Msg(id: Long, msg: Any)
+  case class Response(self: ActorRef)
+  case object StopMessage
+
+  val idExtractor: ShardRegion.ExtractEntityId = {
+    case Msg(id, msg) ⇒ (id.toString, msg)
+  }
+
+  val shardResolver: ShardRegion.ExtractShardId = {
+    case Msg(id, msg) ⇒ (id % 2).toString
+  }
+
+  class PassivatingActor extends Actor with ActorLogging {
+
+    override def preStart(): Unit = {
+      log.info("Starting")
+    }
+
+    override def postStop(): Unit = {
+      log.info("Stopping")
+    }
+
+    override def receive: Receive = {
+      case "passivate" ⇒
+        log.info("Passivating")
+        context.parent ! Passivate(StopMessage)
+        // simulate another message causing a stop before the region sends the stop message
+        // e.g. a persistent actor having a persist failure while processing the next message
+        context.stop(self)
```

Comment:
isn't there a race condition here?
Let's say that the `StopMessage` arrives to the parent BackoffSupervisor before it sees the termination of the child. It will then forward the StopMessage to the dead actor, i.e. lost. That will trigger a backoff and startup the child actor again. Perhaps that outcome is completely alright?

Reply:
Yes you're right.

I see two options. 
* We move to the more specific API you suggested above, that way the back off on stop supervisor knows that it should not restart the client once that predicate has ever been true so once it receives the Terminated it stops its self.
* Have the Shard retry periodically

I think the 1st is the cleanest.

Reply:
Added a test to BackoffSupervisorSpec.scala which catches this case

Reply:
And this stop here catches the case where the child is stopped when receiving the message